clang/lib/CIR/Dialect/Transforms/LifetimeCheck.cpp

```diff
@@ -1653,6 +1995,51 @@ void LifetimeCheckPass::checkNonConstUseOfOwner(mlir::Value ownerAddr,
   return;
 }
 
+void LifetimeCheckPass::markOwnerAsMovedFrom(mlir::Value addr,
+                                             mlir::Location loc) {
+  // Smart pointers have well-defined null state after move, while other
+  // owners are marked invalid.
+  if (isSmartPointerType(addr.getType()))
+    markPsetNull(addr, loc);
+  else
+    markPsetInvalid(addr, InvalidStyle::MovedFrom, loc);
+}
+
+void LifetimeCheckPass::checkMoveCtor(CallOp callOp, cir::CXXCtorAttr ctor) {
+  if (ctor.getCtorKind() != cir::CtorKind::Move)
```

Comment:
Please add a new entry to CIR_CXXSpecialMemberAttr and use the info from FuncOp directly